### DISCUSSION TITLE
Phone authentication in universal app

### DISCUSSION THREAD

**Discussion by MartinN3** [invertase/react-native-firebase#8063] [Q&A] [answered]
I have tackled issue with auth module not working on web, by using firebase sdk standalone web import to get trough authentication. I now need to fill react-native-firebase with auth that i got from other module, is that doable?

**Comment by mikehardy** [comment, marked as answer]
If you are using this on the web, you likely want to shim *all* react-native-firebase APIs to the firebase-js-sdk APIs so that there is no "filling in" and things just work as expected, i.e. https://github.com/invertase/react-native-firebase-authentication-example/tree/main/template/src/shims and the craco.config where webpack is configured with aliases to these shims

**Reply by MartinN3** [reply]
Thank you.

I have read about shim in other answer but haven't connected the dots although I suspected something.

I guess I have to somehow bind this to metro config that handles build instead of webpack. Will try.

I'll just add tags for other developers: I'm exploring tamagui expo solito universal app.

**Reply by MartinN3** [reply]
Works like a charm.

create-tamagui-app with expo router

```js
//babel.config.js
module.exports = (api) => {
  const platform = api.caller(caller => caller && caller.platform);
  api.cache(true)

  return {
    presets: [['babel-preset-expo', { jsxRuntime: 'automatic' }]],
    plugins: [
      platform === 'web' && [
        'module-resolver',
        {
          alias: {
            '@react-native-firebase/app': './shims/firebase-app-web.ts',
            '@react-native-firebase/auth': './shims/firebase-auth-web.ts',
            '@react-native-firebase/firestore': './shims/firebase-firestore-web.ts',
          },
          extensions: ['.ts', '.tsx', '.js', '.json'],
        },
      ],
      [
        '@tamagui/babel-plugin',
        {
          components: ['tamagui'],
          config: './tamagui.config.ts',
          logTimings: true,
          disableExtraction: process.env.NODE_ENV === 'development',
        },
      ],


      // NOTE: this is only necessary if you are using reanimated for animations
      'react-native-reanimated/plugin',
    ].filter(Boolean), // Filter out falsey values to ignore unused plugins,
  }
}
```

```js
//metro.config.js

...
config.resolver.resolveRequest = (context, moduleName, platform) => {
  // Handle Firebase aliases for web
  if (platform === 'web') {
    const aliases = {
      '@react-native-firebase/app$': 'shims/firebase-app-web.ts',
      '@react-native-firebase/auth$': 'shims/firebase-auth-web.ts',
      '@react-native-firebase/firestore$': 'shims/firebase-firestore-web.ts',
    };

    if (aliases[moduleName]) {
      return {
        filePath: aliases[moduleName],
        type: 'sourceFile',
      };
    }
  }

  // Call the default resolver for all other cases
  return context.resolveRequest(context, moduleName, platform);
};
...
```

```js
// _layout.tsx
// in render i added recaptcha container

...
      <Paragraph id="recaptcha-container"></Paragraph>
...

```

```js
//sign-in.tsx
// notice usage of Platform.OS === "web"
...
  const recaptcha = useRef();
  if (Platform.OS === "web") {
    recaptcha.current = new auth.RecaptchaVerifier("recaptcha-container", {
      size: "invisible",
      callback: (response) => {
        // reCAPTCHA solved, allow signInWithPhoneNumber.
      },
    });
  }

async function handlePhoneAuth() {
    if (!loading && confirmationRef) {
      setLoading(true);
      try {
        const result =
          Platform.OS === "web"
            ? await auth().signInWithPhoneNumber(number, recaptcha.current)
            : await auth().signInWithPhoneNumber(number);
        confirmationRef.current = result.confirm.bind(result);
      } catch (e) {
        setLoading(false);
        confirmationRef.current = null;
        const error = e as FirebaseAuthTypes.PhoneAuthError;
        console.log(error);
        // Alert.alert(
        //   appSettings.t("phone-auth-error"),
        //   appSettings.t(error.code ?? "unknownError"),
        //   [{ text: appSettings.t("OK") }]
        // );
      } finally {
        setLoading(false);
      }
    }
  }

...

```

**Reply by mikehardy** [reply]
Glad you got that working! I am using something based off the react-native-firebase-authentication-example myself in production on web/ios/android and it does work well in practice

Just dropping this note to mention that eventually react-native-firebase should support web well as part of platform "other", *and* it should be possible to use the firebase-js-sdk "modular" (also known as v9+) APIs so that tree-shaking in the bundler works etc etc. That *may* even be possible right now I just haven't tested it since "other" platform support went in to beta here so I don't have practical experience (vs the shims, which are a hack, but they do work...)

More info here: https://rnfirebase.io/platforms

**Reply by vipulbhj** [reply]
Hey @mikehardy, I was trying to add a shim for cloud storage, I took inspiration from the firestore implementation and created `firebase-storage-web.ts`.

```typescript
import firebase from 'firebase/compat/app';
import 'firebase/compat/storage';

import initializeApp from './firebase-init';

initializeApp();

const storage = firebase.storage;

// Want to do local development?
// Uncomment this and use `yarn test:emulator:start`
// firestore().useEmulator('http://localhost:8080');

export default storage;
```

This is the original import `import storage, { FirebaseStorageTypes } from '@react-native-firebase/storage';`

I also mapped it to resolve to shim on web, just like described above, but I am seeing an error

```
Unable to resolve "../../../../shims/firebase-app-web.ts/lib/common" from "node_modules/@react-native-firebase/storage/lib/index.js"
```

Any ideas what might be the issue ?

**Reply by mikehardy** [reply]
Nothing specific, but if:

> I also mapped it to resolve to shim on web

...and you see:

> Unable to resolve "../../../../shims/firebase-app-web.ts/lib/common"

...then you have to think that filename it is trying to resolve sure looks wrong doesn't it? Something about your mapping isn't correct, somehow

**Reply by vipulbhj** [reply]
Yeah super weird, don't know why that would be, since I mapped it exactly how I mapped other, let me check this again.

**Reply by vipulbhj** [reply]
@MartinN3 By any chance did you see anything like this ?

**Reply by mikehardy** [reply]
No shims should be needed anymore with other platform (firebase-js-sdk fallback) support

You just need to do an initializeApp step somewhere in your app if `Platform.OS === "web"` before using firebase APIs

https://rnfirebase.io/platforms#other-platforms

**Reply by MartinN3** [reply]
@vipulbhj frustrating experience working with shims, types are somewhat same but on some places different, metro-bundler/babel they often come and go needing to clear everything/resaving/praying (still haven't figured voodoo that works).

Im not sure if related but native authentication vs web authentication, it often stops working out of blue on devices or web. then again restarting reinstalling killing everything seems to be solution.

any idea what it would take to implement missing auth methods for example for phoneNumber? Is there some non obvious obstacle lurking in shadows? I assume it won't be as easy as adding method.

**Reply by mikehardy** [reply]
Huh 🤔  I haven't had much frustration working with shims, and I'm actively coding on a project right now that's using that style (built using the same base code from the rnfb-authentication-example repo I created way back / linked to above)? But I'll believe you, sorry it's not a great experience.

To your question:

> any idea what it would take to implement missing auth methods for example for phoneNumber? Is there some non obvious obstacle lurking in shadows? I assume it won't be as easy as adding method.

I'm not aware of any non-obvious reasons why any of the missing auth methods aren't implemented with the *specific* exception of the multi-factor / TOTP stuff. Unfortunately the firebase emulator suite doesn't handle totp/multi-factor well thus we aren't able to test it well. That's not a hard blocker, but it does make ingesting PRs related to that stuff much more time-consuming as it requires manual verification see https://github.com/firebase/firebase-tools/issues/6224

**Reply by MartinN3** [reply]
Mike thanks for response, it's encouraging.

I was maybe too frustrated with expo, and blamed shims. I must say that shims are reliable, even with emulation. Developing expo app i suggest running web OR apps (not both from one instance). This way it got much more predictable.

Im just scratching surface of firebase emulator authentication and not tried multifactor emulation so no experience there. But with phone auth emulation, it just dumps otp to stdout. Thats problematic to test i suppose, im not sure how i would read it during test in suites i know.

**Reply by mikehardy** [reply]
@MartinN3 we've got you covered since we need to do the same. We sniffed the REST calls made to the emulator and wrote some helpers that can fetch the SMS and OOB codes from the auth emulator and also disable/enable users. This file isn't a public API or anything so I can't guarantee it will stay in one place and we don't ship it with the released packages but you are of course welcome to use whatever's useful, it's open source: https://github.com/invertase/react-native-firebase/blob/main/packages/auth/e2e/helpers.js
